File: simandaya-web/frontend/types/akademik/kategori-kelas.ts
import { UUID } from "../common";

export interface KategoriKelasResponse {
  kategori_kelas_id: UUID;
  tahun_ajaran_id: UUID;
  kode: string;
  nama: string;
  is_active: boolean;
}

export interface KategoriKelasArchiveImpact {
  kategori_kelas_id: UUID;
  kelas_count: number;
  kurikulum_count: number;
}

export interface CreateKategoriKelasRequest {
  tahun_ajaran_id?: UUID;
  kode: string;
  nama: string;
  is_active?: boolean;
}

export interface UpdateKategoriKelasRequest {
  kode?: string;
  nama?: string;
  is_active?: boolean;
}

File: simandaya-web/backend/app/models/kategori_kelas.py
from uuid import UUID, uuid4

from sqlalchemy import Boolean, String, ForeignKey, UniqueConstraint
from sqlalchemy.dialects.postgresql import UUID as SQLAlchemyUUID
from sqlalchemy.orm import Mapped, mapped_column

from app.config.database import Base


class KategoriKelas(Base):
    __tablename__ = "kategori_kelas"
    __table_args__ = (
        UniqueConstraint("tahun_ajaran_id", "kode", name="uq_kategori_tahun_kode"),
        UniqueConstraint("tahun_ajaran_id", "nama", name="uq_kategori_tahun_nama"),
    )

    kategori_kelas_id: Mapped[UUID] = mapped_column(
        SQLAlchemyUUID(as_uuid=True), primary_key=True, default=uuid4, nullable=False
    )
    tahun_ajaran_id: Mapped[UUID] = mapped_column(
        SQLAlchemyUUID(as_uuid=True),
        ForeignKey("tahun_ajaran.tahun_ajaran_id", ondelete="CASCADE"),
        nullable=False,
    )
    kode: Mapped[str] = mapped_column(String(30), nullable=False)
    nama: Mapped[str] = mapped_column(String(100), nullable=False)
    is_active: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)

    def __repr__(self) -> str:
        return f"KategoriKelas(kode={self.kode}, nama={self.nama})"
